README.md
# pixi-easing-graph

A utility component for viewing easing functions in Pixi.js and react-pixi. The graph includes an animated example and can be highly-customized to fit your needs. Function are expected to take a single a number between 0 and 1 and return a number between 0 and 1.

![easing-graph](https://user-images.githubusercontent.com/141928/205437278-2502105e-e92d-4eab-a110-613ac6a5ab9c.png)

## [Demo](mimshwright.github.io/pixi-easing-graph/)

Check out the [Demo](mimshwright.github.io/pixi-easing-graph/) which shows off many of the features and is build with [react-pixi](https://github.com/inlet/react-pixi).

## Usage

```js
import {EasingGraph} from "pixi-easing-graph";

const graph = new EasingGraph(func [, options]);
```

`func` must be an easing function in the format `(x:number) => number` where the range of numbers expected for both input and output are between `0.0` and `1.0`. (see note below about "Drawing Multiple functions")

`options` is an object with the following properties and default values:

- `width` - width of the graph - default `250`
- `height` - height of the graph - default `250`
- `style` - graph style: "dot", "line", or "fill" - default `"dot"`,
- `clamp` - if true, clamps values between the range of 0 and 1 - default `false`
- `steps` - number of steps of the function to draw. If the value is NaN, this will equal `width` - default `NaN`
- `dotSize` - radius of dots when style is "dot" (also used for trails) - default `2`
- `lineWidth` - thickness of line when style is "line" - default `1`
- `background` - background color - default `0xffffff`
- `foreground` - foreground (graph) color - default `0x000000`
- `backgroundAlpha` - alpha for the background color - default `1.0`,
- `fillAlpha` - alpha for the foreground color in "fill" mode - default `0.5`,
- `gridCount` - How many lines in the grid on each axis. If 0, no grid is drawn - default `10`
- `gridSubdivisions` - when true, the middle and edges of the grid are emphasized - default `true`
- `gridColor` - Color of the grid - default `0xcccccc`
- `showMarker` - when true, the animation includes a marker dot that follows the graph - default `true`
- `markerColor` - color of the marker - default `0xff0000`
- `markerSize` - size of the marker - default `10`
- `markerTrail` - when true, the marker leave a trail of dots. If style is `dot` it won't be drawn. - default `false`
- `showMarker` - when true, the animation includes an example animation - default `false`
- `examplePosition` - position of the animated example: "bottom", "right" or "both - default `"bottom"`,
- `exampleColor` - color of the example animation - default `0x333333`
- `exampleSize` - size of the example - default `50`
- `exampleTrail` - when true, the examples leave a trail of dots - default `false`
- `showValues` - when true, shows the input and output values in a small text field - default `false`, fill defaults to `foreground`
- `textStyle` - If `showValues` is `true`, this is the style used to style the text. You can also include `x` and `y` to position the text. Values will be merged with the default. - default `{ fontSize: 10, x: 5, y: 5, fill: foreground]}`
- `duration` - Duration of the animation in milliseconds. - default is `2000`
- `autoPlay` - When true, the animation automatically plays any time `draw()` is called. - default `false`
- `loop` - When true, after you `play()` the animation will loop continuously until you call `stop()` - default `false`

### Animation

To play an animation of the easing function, use the `play()` method.

```js
import { EasingGraph } from "pixi-easing-graph";
const graph = new EasingGraph(quad);
stage.addChild(graph);
graph.play();
```

You can stop animations with `stop()`.

You can adjust the animations by using the `marker-` & `example-` properties in `options`. _More animation options coming soon._

### Drawing Multiple Functions

It's possible to graph more than one function, for exmaple, if you want to compare two easing funcitons. To do this, replace the `func` param in the constructor with an array of functions.

```javascript
const f = (x) => x;
const g = (x) => x * x;
const h = (x) => x * x * x;
const graph = new EasingGraph([f, g, h]);
```

![download](https://user-images.githubusercontent.com/141928/207068303-5e564099-5192-4839-a347-b6d22d4638e0.png)

The first function will be treated as the _primary_ funciton. Only the primary will be animated, show examples, and show text values. All others will only be drawn on the graph. (The ability to draw examples for all functions could be added in the future if needed.)

The functions will be drawn in reverse order so the primary funciton is drawn on top.

By default, the different functions will all use the same styling. To add custom styling to each one, you can create an object for each function...

```javascript
const graph = new EasingGraph([{f:f, foreground: 0x0000FF}, {f:g, foreground:0xFF0000, {f:h, foreground: 0x00FF00}]);
```

Any that don't define `foreground` will use the default value from the options object.

So to summarize, the types that can be passed into the first argument of the constructor are:

```typescript
EasingFunction | (EasingFunction | {f:EasingFunction, foreground?:number })[]
```

## react-pixi

There is a [react-pixi](https://github.com/inlet/react-pixi) component included with the code.

```js
import { EasingGraphComponent as EasingGraph } from "pixi-easing-graph";
```

All the options can be passed to this component as props as well as `f`, `x`, `y`.

As a hack, to trigger the animation, you can also set the `play` property to a different value to trigger the animation.

```js
<button onClick={()=> setPlay(play === 0 ? 1 : 0)}>
```

## Development

### Scripts

- `dev` : start a server to see the demo
- `lint` : lint the code. You can also do `fix` to auto-fix.
- `test` : test with jest. Also `test:coverage` and `test:watch`
- `build` : bundle the code
- `commit` & `release` : see below
- `deploy`: Build the demo code and publish to gh-pages

### Commits & Releases

Code is automatically linted before being committed. I recommend installing the plugins for eslint and prettier in your code editor. You can attempt to fix linting issues with `yarn fix`.

When ready to commit, please commit using `yarn commit` to use commitizen for standard format commits.

When ready to release use `yarn release` with the `-r patch|minor|major` flag (default without the flag is `patch`).

You'll then need to publish your changes separately.

You can deploy an updated demo to github pages using `yarn deploy`


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.1.5](https://github.com/mimshwright/pixi-easing-graph/compare/v1.1.4...v1.1.5) (2023-01-28)

### [1.1.4](https://github.com/mimshwright/pixi-easing-graph/compare/v1.1.3...v1.1.4) (2022-12-14)

### Bug Fixes

- **cleanup:** fixed an issue where animations weren't stopped before removing from render ([74f3459](https://github.com/mimshwright/pixi-easing-graph/commit/74f3459e5bebde2d6d16943715281a60d8da477e))

### [1.1.3](https://github.com/mimshwright/pixi-easing-graph/compare/v1.1.2...v1.1.3) (2022-12-14)

### Bug Fixes

- **animation:** fixed a bug with looping animaitons ([47beb3e](https://github.com/mimshwright/pixi-easing-graph/commit/47beb3ea36373cc7d9fc5161aa55c9de29f821b6))

### [1.1.2](https://github.com/mimshwright/pixi-easing-graph/compare/v1.1.1...v1.1.2) (2022-12-13)

### [1.1.1](https://github.com/mimshwright/pixi-easing-graph/compare/v1.1.0...v1.1.1) (2022-12-12)

## [1.1.0](https://github.com/mimshwright/pixi-easing-graph/compare/v1.0.0...v1.1.0) (2022-12-12)

### Features

- **multiple functions:** added the ability to display multiple functions in the graph ([8392119](https://github.com/mimshwright/pixi-easing-graph/commit/8392119eb6bd4516d7ab15177544a27caac3d692))
- **text:** added ability to style the text field ([5729394](https://github.com/mimshwright/pixi-easing-graph/commit/5729394f92cd9e4997e53f060b871b691b0eba54))

## [1.0.0](https://github.com/mimshwright/pixi-easing-graph/compare/v0.1.6...v1.0.0) (2022-12-11)

### Features

- **styles, text:** impoved styling of the demo significantly. Added additional options to graph ([8ec12f5](https://github.com/mimshwright/pixi-easing-graph/commit/8ec12f524122bde487cb6c47719259cf25218f71))

### [0.1.6](https://github.com/mimshwright/pixi-easing-graph/compare/v0.1.5...v0.1.6) (2022-12-10)

### Bug Fixes

- **darkmode:** changed the way dark mode is detected to happen on redraw instead of page refresh ([87509c6](https://github.com/mimshwright/pixi-easing-graph/commit/87509c6ed77916643c2d4e65280ee3fdca74c4a0))

### [0.1.4](https://github.com/mimshwright/pixi-easing-graph/compare/v0.1.3...v0.1.4) (2022-12-05)

### Features

- **style:** added backgroundAlpha property ([f5465e0](https://github.com/mimshwright/pixi-easing-graph/commit/f5465e01556d6161179e0ce57f27c6daa428876c))

### [0.1.3](https://github.com/mimshwright/pixi-easing-graph/compare/v0.1.2...v0.1.3) (2022-12-03)

### Bug Fixes

- **build:** fixed issues with gh-pages (for real this time) ([37b679d](https://github.com/mimshwright/pixi-easing-graph/commit/37b679db4a97fd7534ab93a3ff2f9b17f65f495f))

### [0.1.2](https://github.com/mimshwright/pixi-easing-graph/compare/v0.1.1...v0.1.2) (2022-12-03)

### Bug Fixes

- **demo:** fixed the build issues with the demo and integrated with gh pages ([760bf80](https://github.com/mimshwright/pixi-easing-graph/commit/760bf80905ed5d259361b7a82dc7c0e9052dfef6))

### 0.1.1 (2022-12-03)

### Features

- **animation:** added options for controlling animations ([d15ed40](https://github.com/mimshwright/pixi-easing-graph/commit/d15ed40af765714fbb45d7651f3af241c1e5daf5))
- **init:** initial commit! ([b831375](https://github.com/mimshwright/pixi-easing-graph/commit/b83137520cefde20e7307d5473cd53db57ba2130))

# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.


src/EasingGraph.ts
import { Sprite, Graphics, Texture, Ticker, Text, TextStyle } from "pixi.js";
import {
  apply,
  times,
  zip,
  clamp,
  identity,
  defaultTo,
  prop,
  pipe,
  reverse,
} from "ramda";

type Point = [number, number];
export type EasingFunction = (x: number) => number;

export type EasingGraphStyle = "dot" | "line" | "fill";
export type ExamplePosition = "bottom" | "right" | "both";

export interface EasingFunctionOptions {
  f: EasingFunction;
  foreground?: number;
  markerColor?: number;
}
type FuncOrFuncs = EasingFunction | (EasingFunction | EasingFunctionOptions)[];
export interface EasingGraphOptions {
  width: number;
  height: number;
  steps: number;
  style: EasingGraphStyle;
  clamp: boolean;
  background: number;
  foreground: number;
  fillAlpha: number;
  backgroundAlpha: number;
  showMarker: boolean;
  markerColor: number;
  markerSize: number;
  markerTrail: boolean;
  showExample: boolean;
  exampleColor: number;
  exampleSize: number;
  examplePosition: ExamplePosition;
  exampleTrail: boolean;
  dotSize: number;
  lineWidth: number;
  showValues: boolean;
  textStyle: TextStyle & { x?: number; y?: number };
  gridColor: number;
  gridCount: number;
  gridSubdivisions: boolean;
  duration: number;
  autoPlay: boolean;
  loop: boolean;
}

const ticker = Ticker.shared;

const defaultTextStyle = { fontSize: 10, x: 5, y: 5 };

const defaultOptions: EasingGraphOptions = {
  width: 250,
  height: 250,
  style: "dot",
  clamp: false,
  steps: NaN,
  dotSize: 2,
  lineWidth: 1,
  background: 0xffffff,
  foreground: 0x000000,
  backgroundAlpha: 1.0,
  fillAlpha: 0.5,
  showMarker: true,
  markerColor: 0xff0000,
  markerSize: 10,
  markerTrail: false,
  showExample: false,
  exampleColor: 0x333333,
  exampleSize: 50,
  examplePosition: "bottom",
  exampleTrail: false,
  showValues: false,
  textStyle: {} as TextStyle,
  gridCount: 10,
  gridColor: 0xcccccc,
  gridSubdivisions: true,
  duration: 2000,
  autoPlay: false,
  loop: false,
};

const clamp01 = clamp(0, 1);
const fixed = (d: number) => (x: number) => Math.round(x * 10 ** d) / 10 ** d;

class EasingGraph extends Sprite {
  _f: EasingFunctionOptions[] = [];
  set f(f: FuncOrFuncs) {
    // Convert type of EasingFunction | (EasingFunction | EasingFunctionOptions)[] to
    // EasingFunctionOptions[];
    this._f =
      f instanceof Array
        ? f.map((f) => (typeof f === "function" ? { f } : f))
        : [{ f }];
    if (this._f.length === 0) {
      throw new Error(
        "There must be at least one easing function attached to the graph."
      );
    }
  }
  get f(): EasingFunctionOptions[] {
    return this._f;
  }

  graphics: Graphics;
  trail: Graphics;
  options: EasingGraphOptions;

  isPlaying = false;
  marker: Marker;
  exampleX: Marker;
  exampleY: Marker;
  text: Text;
  t = 0;

  static defaultOptions = defaultOptions;

  constructor(f: FuncOrFuncs, options: Partial<EasingGraphOptions> = {}) {
    super();

    this.f = f;

    this.options = { ...defaultOptions, ...options };

    this.graphics = new Graphics();
    this.addChild(this.graphics);

    this.trail = new Graphics();
    this.addChild(this.trail);

    const { markerSize, markerColor, exampleColor, exampleSize } = this.options;

    this.marker = new Marker(markerColor, markerSize);
    this.marker.visible = false;
    this.addChild(this.marker);

    this.exampleX = new Marker(exampleColor, exampleSize);
    this.exampleX.visible = false;
    this.exampleX.x = 0;
    this.exampleX.y = this.options.height + exampleSize * 1.5;
    this.addChild(this.exampleX);

    this.exampleY = new Marker(exampleColor, exampleSize);
    this.exampleY.visible = false;
    this.exampleY.x = this.options.width + exampleSize * 1.5;
    this.exampleY.y = this.options.height;
    this.addChild(this.exampleY);

    const mergedStyle = {
      ...{ fill: options.foreground },
      ...defaultTextStyle,
      ...options.textStyle,
    };

    this.text = new Text("", mergedStyle);
    this.text.visible = false;
    this.text.x = mergedStyle.x;
    this.text.y = mergedStyle.y;
    this.addChild(this.text);
  }

  play() {
    this.stop();

    const { showMarker, showExample, examplePosition, showValues } =
      this.options;

    this.marker.visible = showMarker;
    this.exampleY.visible = this.exampleX.visible = showExample;
    this.text.visible = showValues;

    if (examplePosition === "bottom") this.exampleY.visible = false;
    if (examplePosition === "right") this.exampleX.visible = false;

    this.trail.clear();
    this.t = 0;

    this.isPlaying = true;
    ticker.add<EasingGraph>(this.animationStep, this);
  }
  stop() {
    ticker.remove<EasingGraph>(this.animationStep, this);
    this.isPlaying = false;
  }

  private updateText(x: number, y: number) {
    const decimals = fixed(3);
    this.text.text = `x: ${decimals(x)}\ny: ${decimals(y)}`;
  }

  private animationStep() {
    this.t += ticker.deltaMS;
    const { options, exampleX: ex, exampleY: ey, marker, trail } = this;
    const {
      clamp,
      width,
      height,
      exampleTrail,
      markerTrail,
      dotSize,
      foreground,
      duration,
      loop,
      style,
      showValues,
    } = options;

    this.t = Math.min(this.t, duration);
    const t = this.t;

    const clampFunction = clamp ? clamp01 : identity;

    const f = this.f[0].f;
    const x = t / duration;
    const y = clampFunction(f(x));

    marker.x = x * width;
    marker.y = height - y * height;

    if (ex.visible) ex.x = y * width;
    if (ey.visible) ey.y = marker.y;

    const g = trail;
    g.beginFill(foreground);
    if (exampleTrail) {
      if (ex.visible) g.drawCircle(ex.x, ex.y, dotSize);
      if (ey.visible) g.drawCircle(ey.x, ey.y, dotSize);
    }
    if (markerTrail && style !== "dot") {
      if (marker.visible) g.drawCircle(marker.x, marker.y, dotSize);
    }
    g.endFill();

    if (showValues) {
      this.updateText(x, y);
    }

    if (t >= duration) {
      this.stop();
      if (loop) {
        this.play();
        this.draw();
      }
    }
  }
  draw() {
    const g = this.graphics;
    const {
      width,
      height,
      style,
      foreground,
      background,
      backgroundAlpha,
      steps: stepsOrNaN,
      gridCount,
      clamp,
      autoPlay,
    } = this.options;
    const steps = isNaN(stepsOrNaN) ? width : stepsOrNaN;

    const clampFunction = clamp ? clamp01 : identity;

    // reverse so that the last functions are drawn first.
    const fs = reverse(this.f);
    const inputs = times((n) => n / (steps - 1), steps);
    const outputs = fs.map((f: EasingFunctionOptions) =>
      inputs
        .map(f.f)
        // clamp values?
        .map(clampFunction)
    );
    const coords: Point[][] = outputs.map((fOutputs) => zip(inputs, fOutputs));
    const coordToPixel = ([x, y]: Point): Point => [
      x * width,
      height - y * height,
    ];
    const pixelCoords: Point[][] = coords.map((fCoords) =>
      fCoords.map(coordToPixel)
    );
    const foregroundColors = fs.map(
      pipe(prop("foreground"), defaultTo(foreground))
    );
    const drawParams = zip(pixelCoords, foregroundColors);

    g.clear();
    g.beginFill(background, backgroundAlpha);
    g.drawRect(0, 0, width, height);
    g.endFill();

    if (gridCount > 0) {
      this.drawGrid();
    }

    const drawFunctions = {
      line: this.drawLines,
      fill: this.drawFill,
      dot: this.drawDots,
    };
    drawParams.map((params) => drawFunctions[style].apply(this, params));

    if (autoPlay && this.isPlaying === false) {
      this.play();
    }
  }

  drawDots(coords: Point[], foreground: number) {
    const { dotSize } = this.options;
    const g = this.graphics;
    const drawDot = (x: number, y: number) => g.drawCircle(x, y, dotSize);

    g.beginFill(foreground);
    coords.map(apply(drawDot));
    g.endFill();
  }
  drawLines(coords: Point[], foreground: number) {
    const { height, lineWidth } = this.options;
    const g = this.graphics;
    const drawLine = apply(g.lineTo.bind(g));

    g.moveTo(0, height);
    g.lineStyle(lineWidth, Math.floor(foreground));
    coords.map(drawLine);
    g.lineStyle();
  }
  drawFill(coords: Point[], foreground: number) {
    const { fillAlpha, width, height } = this.options;
    const g = this.graphics;
    const drawLine = apply(g.lineTo.bind(g));

    g.moveTo(0, height);
    g.beginFill(foreground, fillAlpha);
    coords.map(drawLine);
    g.lineTo(width, 0);
    g.lineTo(width, height);
    g.lineTo(0, height);
    g.endFill();
    g.closePath();
  }

  drawGrid() {
    const g = this.graphics;
    const { width, height, gridCount, gridColor, gridSubdivisions } =
      this.options;

    g.lineStyle(1, gridColor);
    const drawGridLines = (pos: number) => {
      g.lineStyle(1, gridColor);
      if (
        (gridSubdivisions && pos === 0) ||
        pos === gridCount ||
        (pos === gridCount / 2 && gridCount % 2 === 0)
      ) {
        g.lineStyle(2, gridColor);
        g.lineStyle(2, gridColor);
      }

      g.moveTo(0, (pos / gridCount) * height);
      g.lineTo(width, (pos / gridCount) * height);
      g.closePath();
      g.moveTo((pos / gridCount) * width, 0);
      g.lineTo((pos / gridCount) * width, height);
      g.closePath();
    };
    times(drawGridLines, gridCount + 1);
    g.lineStyle();
  }
}
export default EasingGraph;

class Marker extends Sprite {
  color: number;
  size: number;
  graphics: Graphics;

  constructor(color: number, size: number, texture?: Texture) {
    super(texture);
    this.color = color;
    this.size = size;
    this.graphics = new Graphics();
    this.addChild(this.graphics);
    this.anchor.set(0.5);

    this.draw();
  }
  draw() {
    const { graphics: g, color, size } = this;
    g.clear();
    g.beginFill(color);
    g.drawCircle(0, 0, size);
    g.endFill();
  }
}
